### ISSUE TITLE
Upgrade to i18next v2.0.0

### ISSUE THREAD

**Issue by Mordred** [aurelia/i18n#46]
I18next is now [v2.0.0](https://github.com/i18next/i18next/releases/tag/2.0.0). It would be nice to able to upgrade.

**Comment by zewa666** [comment]
Are there any breaking changes you know of which could cause issues?

**Comment by Mordred** [comment]
If you switch [setup](https://github.com/aurelia/i18n/blob/master/src/i18n.js#L26) to new version it should be OK. See http://i18next.com/docs/migration/

**Comment by heruan** [comment]
:+1: Absolutely, [v2.0.0](https://github.com/i18next/i18next/releases/tag/2.0.0) has full ES6 support, letting import/use plugins with:
```javascript
import Cache from 'i18next-localstorage-cache';

i18next.use(Cache);
```
See http://i18next.com/docs/

**Comment by zewa666** [comment]
hmmm there are several issues with updating to the new version. Not only there are massive API changes, but also the separation of loading mechanisms and move to plugins makes the way we used to work in this plugin totally different.

In order to fully support each option, either the Aurelia user has to install more things on his own and do additional settings, or I preload all of them, which on the other hand is not the most efficient option.

So I've started a separate branch to take a look at it but found that it's currently a too massive change which I'm not yet sure if its worth the pain. 

What I would like to find out in order to continue with the upgrade is:

- Do we have any new real benefits, besides ES6
- How do we go ahead with the lacking default backend (e.g. use XHR as default?!?)
- Do we include caching out of the box?
- Do we need more facades to the original API?

It's certainly doable, just not a thing of few minutes :)

**Comment by Mordred** [comment]
What about dropping dependency on i18next and letting users configure it during bootstrap? E.g.

```js
import i18next from 'i18next';
import XHR from 'i18next-xhr-backend';
import Cache from 'i18next-localstorage-cache';
import LanguageDetector from 'i18next-browser-languagedetector';
import sprintf from 'i18next-sprintf-postprocessor';

i18next
  .use(XHR) // or any other backend implementation
  .use(Cache) // or any other cache implementation
  .use(LanguageDetector) // or any other implementation
  .use(sprintf) // or any other post processor
  .init(options);

aurelia.use.plugin('aurelia-i18n', (i18n) => {
  i18n.setProvider(i18next);
})
```

All you have to do then is provide some kind of interface which should be implemented for the provider.

**Comment by heruan** [comment]
I think that the upgrade is necessary, Aurelia is still in beta and it "leverages the cutting edge capabilities of the web" and having an official (and great!) plugin going stable with an "old" dependency doesn't fit very well.

Said that, the plugin must be self-sufficient with a small set of feature, i.e. a simple `aurelia.use.plugin('aurelia-i18n')` must suffice to have the plugin working with a minimum set of requirements/dependencies. For special needs, the configuration callback will let the advanced user to configure the `i18next` instance with the new `i18next.use(…).init(options)`.

As @zewa666 said, It's not a thing of few minutes, but definitely worth them.

**Comment by zewa666** [comment]
> Said that, the plugin must be self-sufficient with a small set of feature, i.e. a simple aurelia.use.plugin('aurelia-i18n') must suffice to have the plugin working with a minimum set of requirements/dependencies

@heruan thats exactly what I was thinking of. The approach @Mordred presents is cool but introduces the following facts.

Aurelia-i18n becomes an abstraction layer for i18n. But i18next itself became exactly that in v2. So are we starting to over-abstract things?

And secondly if the plugin is too hard to configure, it will be harder to use. That may lead to people not using it at all because it looks too hard.

So that's why I'd like to take a sensitive approach at the begin and offer something "out of the box" but also provide facades for important configurations.

**Comment by zewa666** [comment]
@heruan and @Mordred I've created a new branch targeting i18next v2. We now should have no installing issues with neither JSPM nor NPM. The current version does not support custom file loading but only providing translations via init.options parameter.

So I'd like to ask you guys, or anybody else to give this version a test run and see whether this fits the new needs. In the meantime I'll see how to go about the default loader

**Comment by Mordred** [comment]
I have tested your branch and everything seems to be OK

**Comment by zewa666** [comment]
oh cool thanks for the hint, funny enough though it seems to work on the root declaration as well. nevertheless it shall be moved :)

**Comment by AdamWillden** [comment]
@zewa666 I'm completely new to using i18n and this plugin. I set up using the aforementioned branch yesterday with no issues so far :+1: 

At the moment however I'm only doing simple translations with no date or number formatting and no gender or pluralisation handling. I am using namespaces already.

When I progress to using number and date formatting I'll try to give more feedback if still necessary.

Many thanks for the plugin.

**Comment by AdamWillden** [comment]
@zewa666 You need to update the documentation to reflect the changes to interpolation (parameters) from `__blah__` to `{{blah}}`. I've noticed that you put a note in the migration tips but didn't update the main docs.

So, as you've guessed I've just done string interpolation which also contained nested translations. That all worked okay.  Have done some dateTime formatting which worked perfectly also.

Not hit any issues yet.

**Comment by zewa666** [comment]
ok thats a good point, thanks for reviewing :+1:

**Comment by Jenselme** [comment]
Just completed my tests of the new version with webpack. Here's what I found in addition to the remarks of @AdamWillden:

- I didn't manage to get the current release working with webpack. With i18next 2.3.X it works without problems. Just check https://github.com/i18next/i18next-xhr-backend#usage-with-webpack to know how to load the translations files
- I encountered a problem while changing translation by code (although I suspect this is a bug in i18next). I have two languages: fr and en and one fallback language en. If I define the language to be en when the application starts, I cannot change to fr. But if I define fr, I can switch to en. I solve this, by using an array of fallback languages: `['fr', 'en']`.

**Comment by zewa666** [comment]
@Jenselme thanks for the tests. It would be great if you could test the later issue in an i18next only environment to see whether it's a i18next related issue or if its on our side.

**Comment by AdamWillden** [comment]
@zewa666 @EisenbergEffect is this due to be closed now following the [0.5.0 release announcement](http://blog.durandal.io/2016/03/25/i18n-update-release/)? Or have I misunderstood somehow. Also am I correct that the interpolation is still `__ __` rather than `{{ }}`

**Comment by zewa666** [comment]
nope, the standard interpolation did change to {{blah}}. Take a look at the update notes https://github.com/aurelia/i18n#migration-to-new-i18next-version

**Comment by AdamWillden** [comment]
Oh okay It was something I saw in the commit that led me to believe otherwise. I must have misunderstood.  Thanks

**Comment by Jenselme** [comment]
@zewa666 I cannot reproduce in a very basic i18next app but I can in an Aurelia one.

- [Here is my sample Aurelia app](http://jenselme.perso.centrale-marseille.fr/visible/skeleton-es2016-webpack-i18n.zip) (relevant code in `src/main.js` and `src/welcome.js`).
- [here is my test with i18next](http://jenselme.perso.centrale-marseille.fr/visible/i18next.zip).

What do you think?

**Comment by Jenselme** [comment]
> I encountered a problem while changing translation by code (although I suspect this is a bug in i18next). I have two languages: fr and en and one fallback language en. If I define the language to be en when the application starts, I cannot change to fr. But if I define fr, I can switch to en. I solve this, by using an array of fallback languages: ['fr', 'en'].

@zewa666 For your information, I think it comes from this plugin. The solution I implemented in #92 solved it (among other things).
